flower/utils/pg_storage.py

```diff
@@ -27,7 +27,14 @@
 
 _add_event = """INSERT INTO events (time, data) VALUES (%s, %s) ON CONFLICT DO NOTHING"""
 
-_all_events = """SELECT data FROM events ORDER BY time ASC"""
+_get_events = """SELECT data FROM (
```

Comment:
@johnarnold Limiting by a number events can result in an incomplete task load at the point where loading is cut off. I've just tested this query for loading the events for the last x tasks:

```sql
SELECT data FROM events WHERE data->'uuid' IN (
  SELECT data->'uuid' FROM events WHERE data @> '{"type": "task-received"}' ORDER BY time DESC LIMIT 4
)
ORDER BY time ASC
```

Reply:
max_events should never be none, as it's defaulted to 30000 (i.e. 3x the default max_tasks)

Also, it's loading the newest events, so while you could get a partial task load, it will be the oldest tasks.  Does that cause any problems?  I don't see a way to avoid it, other than keeping 'all events forever' and that doesn't seem doable or useful either.

Reply:
My thought was that if you load enough or slightly more events than are needed to populate the task cache up to it's max_tasks size, any partial tasks will get dropped out of the cache

Reply:
I have this running in production, btw.  My ratios might be slightly off --
storage_max_events = 300k  yielded ~ 75k tasks, whereas my max_tasks = 100k.

but it does work.

Reply:
@johnarnold partially loaded tasks would be an issue when looking at *active tasks* (the `/tasks?state=STARTED` endpoint, linked to from the dashboard). Regardless of it being old, a task would likely show up there if not all events for it have been processed.

It can't be stated with 100% accuracy how many events there will be per task, and relying on the LRU cache fixing things prevents you from choosing to load a small number of events.

You haven't commented on the query I've suggested that loads events for the last x tasks. That would avoid partially loading tasks.

Also, there is no reason why there should not remain the option to load all events -- just leave in the previous query that does this and execute is if max_events is None.

Reply:
@bosonogi sorry, I didn't understand the nuance of your query on first read. Need more coffee.

It's clever, let me integrate the change and re-test.

Reply:
Using your query, we could re-use max_tasks option instead of storage_max_events.  That would work well for postgresql assuming the load performance is ok.   For other storage providers (e.g. redis) it may be more straightforward to use storage_max_events and just eat the small number of partial task loads.

Reply:
A quick test shows LIMIT 3000 takes 6mins to query, with a database of ~3M events. So we may need to rethink the query for performance, as that seems like a long time and the databases will just grow and grow. Need to further constrain the query.

Reply:
I'm not sure a contains-type query against the GIN index will get us the performance we want.  I'm not a pSQL expert so maybe there's a query plan optimization we're missing.  My head goes towards making uuid it's own column and indexing it, so we can do a simpler distinct + order + limit query to build a list of n tasks, and fetch all their events.

Reply:
@johnarnold Another approach might be to keep the query simple as it was, and move the limit logic to Python code. As we read all events from the database one at a time in descending order, we could detect when all events for `max_tasks` number of tasks have been loaded and stop reading. Then these events could be replayed in the correct (ascending) order.

This would involve loading the events to memory before using them, but by design that amount of data must be small enough to fit temporarily in memory without issues.

Reply:
@bosonogi That actually could be pretty straightforward, as a simple Set of uuids with a len() check against max_tasks would tell us when to stop. I'm a bit worried though that the 'get all events' query will be slow to return a cursor as the table grows.  I'll do some testing.   I think exposing uuid as a column may end up giving the desired perf and keep the query on the simple side.

As an aside, I'm seeing utilization on my postgresql server growing at a rate of about 1% per-hour until it flatlines (70% currently) -- I suspect the GIN index becomes troublesome to update with a very large table size.   So we should either think about occasionally truncating the table, or ditching the GIN index.  Give it some thought, and let me know if you see similar behavior?

Reply:
Interesting reading: ORDER BY kills gin index perf, forces b-tree index:
https://dba.stackexchange.com/questions/113132/gin-index-not-used-when-adding-order-clause

Reply:
Interesting reading indeed. We do order by on the "time" column, so we might get away with it.

I'm not against adding one more column and index for UUID if it proves that would produce significant boost in performance. Note however that the UUID field is unfortunately named as such as it is possible to define any string as the task ID (and we do that in our system so that the task IDs would be more meaningful). This means you can't limit the column type with assumptions related to UUIDs.